File: packages/android/storage-android/src/main/kotlin/app/notes/storage/GoogleDriveAdapter.kt
package app.notes.storage

import app.notes.domain.FileInput
import app.notes.domain.RemoteFile
import app.notes.domain.SIMPLE_UPLOAD_THRESHOLD_BYTES
import app.notes.domain.StorageError
import app.notes.domain.StorageException
import app.notes.domain.StorageProvider
import app.notes.domain.StorageProviderId
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

private const val APP_FOLDER_NAME = "NotesApp"

/**
 * V1 storage adapter. Wraps the Drive v3 REST surface via the [DriveHttp]
 * port. Auth-agnostic: the constructor takes a [TokenProvider] that asks
 * the auth module for a fresh token on every call.
 */
class GoogleDriveAdapter(
    private val http: DriveHttp,
    private val tokenProvider: TokenProvider,
    private val backoff: BackoffConfig = BackoffConfig(),
) : StorageProvider {
    override val id: StorageProviderId = StorageProviderId.GoogleDrive

    @Volatile private var folderId: String? = null
    private val folderMutex = Mutex()

    override suspend fun ensureAppFolder(): String {
        folderId?.let { return it }
        folderMutex.withLock {
            folderId?.let { return it }
            val resolved = lookupOrCreateFolder()
            folderId = resolved
            return resolved
        }
    }

    override suspend fun upload(file: FileInput, folderHint: String?): RemoteFile {
        val folder = ensureAppFolder()
        return if (file.sizeBytes < SIMPLE_UPLOAD_THRESHOLD_BYTES) {
            simpleUpload(folder, file, folderHint)
        } else {
            resumableUpload(folder, file, folderHint)
        }
    }

    override suspend fun getDownloadUrl(externalId: String, ttlSeconds: Int): String {
        if (externalId.isBlank()) {
            throw StorageException(StorageError.NotFound("externalId is required"))
        }
        if (ttlSeconds !in 1..3600) {
            throw StorageException(StorageError.TransportError("ttlSeconds must be in (0, 3600]"))
        }
        val token = requireToken()
        callWithRetry { http.getMetadata(token, externalId) }
        return "https://www.googleapis.com/drive/v3/files/" +
            externalId.encodeUrl() +
            "?alt=media&access_token=" +
            token.encodeUrl()
    }

    override suspend fun delete(externalId: String) {
        val token = requireToken()
        callWithRetry { http.delete(token, externalId) }
    }

    private suspend fun simpleUpload(
        folder: String,
        file: FileInput,
        folderHint: String?,
    ): RemoteFile {
        val bytes = file.bytes
            ?: throw StorageException(StorageError.TransportError("simpleUpload requires FileInput.bytes"))
        val token = requireToken()
        val meta = callWithRetry {
            http.simpleUpload(token, folder, file.name, file.mimeType, bytes, folderHint)
        }
        return meta.toRemoteFile()
    }

    private suspend fun resumableUpload(
        folder: String,
        file: FileInput,
        folderHint: String?,
    ): RemoteFile {
        val streamFactory = file.stream
            ?: throw StorageException(StorageError.TransportError("resumableUpload requires FileInput.stream"))
        val token = requireToken()
        val session = callWithRetry {
            http.startResumableUpload(
                token,
                folder,
                file.name,
                file.mimeType,
                file.sizeBytes,
                folderHint,
            )
        }
        val chunks = InputStreamChunkSource(streamFactory())
        var offset = 0L
        var finalMeta: DriveFileMeta? = null
        while (true) {
            val chunk = chunks.nextChunk() ?: break
            if (chunk.isEmpty()) continue
            finalMeta = callWithRetry {
                http.putChunk(session.sessionUri, chunk, offset, file.sizeBytes)
            }
            offset += chunk.size
        }
        if (offset != file.sizeBytes) {
            throw StorageException(
                StorageError.TransportError(
                    "Resumable upload truncated: sent $offset of ${file.sizeBytes} bytes",
                ),
            )
        }
        val resolved = finalMeta
            ?: throw StorageException(
                StorageError.UpstreamError(500, "Resumable upload completed without final metadata"),
            )
        return resolved.toRemoteFile()
    }

    private suspend fun lookupOrCreateFolder(): String {
        val token = requireToken()
        val existing = callWithRetry { http.findAppFolder(token, null, APP_FOLDER_NAME) }
        if (existing != null) return existing
        return callWithRetry { http.createFolder(token, null, APP_FOLDER_NAME) }
    }

    private suspend fun requireToken(): String {
        val t = tokenProvider()
        if (t.isNullOrEmpty()) {
            throw StorageException(
                StorageError.Unauthorized("No Google Drive token available; ask the user to re-link"),
            )
        }
        return t
    }

    private suspend fun <T> callWithRetry(op: suspend () -> T): T =
        retry(op, ::isRetriable, backoff)
}

private fun DriveFileMeta.toRemoteFile(): RemoteFile = RemoteFile(
    externalId = id,
    name = name,
    mimeType = mimeType,
    sizeBytes = size,
)

internal fun isRetriable(err: Throwable): Boolean {
    if (err !is StorageException) return false
    return when (val inner = err.error) {
        is StorageError.TransportError -> true
        is StorageError.UpstreamError -> inner.status == 408 || inner.status == 429 || inner.status >= 500
        else -> false
    }
}

private fun String.encodeUrl(): String =
    java.net.URLEncoder.encode(this, java.nio.charset.StandardCharsets.UTF_8)

File: packages/storage/src/google-drive-adapter.ts
import {
  SIMPLE_UPLOAD_THRESHOLD_BYTES,
  StorageError,
  StorageException,
  type FileInput,
  type RemoteFile,
  type StorageProvider,
  type StorageProviderId,
} from "../../domain/src/index.ts";
import { DEFAULT_BACKOFF, retry, type BackoffConfig } from "./backoff.js";
import type { DriveHttpClient, DriveListMatch } from "./drive-http.js";

const APP_FOLDER_NAME = "NotesApp";
/** Drive's recommended chunk size; must be a multiple of 256 KB. */
const RESUMABLE_CHUNK_BYTES = 8 * 1024 * 1024;

/** Tokens supplied by the auth module on every call. */
export type TokenProvider = () => Promise<string | null>;

export interface GoogleDriveAdapterOptions {
  readonly http: DriveHttpClient;
  readonly token: TokenProvider;
  readonly backoff?: BackoffConfig;
}

/**
 * V1 storage adapter. Talks to Drive v3 via the [DriveHttpClient] port —
 * never imports the heavy `googleapis` package, keeping bundle size
 * reasonable.
 *
 * Folder caching: [ensureAppFolder] does a search on first call, creates
 * the folder if missing, and caches the id for the lifetime of the
 * adapter. Re-instantiate to force a re-lookup (e.g. on user switch).
 */
export class GoogleDriveAdapter implements StorageProvider {
  readonly id: StorageProviderId = "GOOGLE_DRIVE";
  private folderId: string | null = null;
  private folderPromise: Promise<string> | null = null;
  private readonly backoff: BackoffConfig;

  constructor(private readonly options: GoogleDriveAdapterOptions) {
    this.backoff = options.backoff ?? DEFAULT_BACKOFF;
  }

  async ensureAppFolder(): Promise<string> {
    if (this.folderId !== null) return this.folderId;
    if (this.folderPromise !== null) return this.folderPromise;
    this.folderPromise = this.lookupOrCreateFolder();
    try {
      this.folderId = await this.folderPromise;
      return this.folderId;
    } finally {
      this.folderPromise = null;
    }
  }

  async upload(file: FileInput, folderHint?: string): Promise<RemoteFile> {
    const folderId = await this.ensureAppFolder();
    if (file.sizeBytes < SIMPLE_UPLOAD_THRESHOLD_BYTES) {
      return this.simpleUpload(folderId, file, folderHint);
    }
    return this.resumableUpload(folderId, file, folderHint);
  }

  async getDownloadUrl(externalId: string, ttlSeconds = 300): Promise<string> {
    // Drive doesn't expose pre-signed URLs the way S3 does; the canonical
    // pattern is to embed a short-lived access token in the URL. The
    // storage/sign Edge Function is the source of truth — see storage-sign.ts.
    if (externalId.trim().length === 0) {
      throw new StorageException(StorageError.notFound("externalId is required"));
    }
    if (ttlSeconds <= 0 || ttlSeconds > 3600) {
      throw new StorageException(StorageError.transport("ttlSeconds must be in (0, 3600]"));
    }
    const token = await this.requireToken();
    // Sanity check the file exists; lets us surface NotFound before the
    // browser hits a 404.
    await this.callWithRetry(() => this.options.http.getMetadata(token, externalId));
    return `https://www.googleapis.com/drive/v3/files/${encodeURIComponent(
      externalId,
    )}?alt=media&access_token=${encodeURIComponent(token)}`;
  }

  async delete(externalId: string): Promise<void> {
    const token = await this.requireToken();
    await this.callWithRetry(() => this.options.http.delete(token, externalId));
  }

  private async simpleUpload(
    folderId: string,
    file: FileInput,
    folderHint: string | undefined,
  ): Promise<RemoteFile> {
    if (file.bytes === undefined) {
      throw new StorageException(StorageError.transport("simpleUpload requires file.bytes"));
    }
    const token = await this.requireToken();
    const result = await this.callWithRetry(() =>
      this.options.http.simpleUpload(
        token,
        folderId,
        file.name,
        file.mimeType,
        file.bytes!,
        folderHint,
      ),
    );
    return matchToRemoteFile(result);
  }

  private async resumableUpload(
    folderId: string,
    file: FileInput,
    folderHint: string | undefined,
  ): Promise<RemoteFile> {
    if (file.stream === undefined) {
      throw new StorageException(StorageError.transport("resumableUpload requires file.stream"));
    }
    const token = await this.requireToken();
    const session = await this.callWithRetry(() =>
      this.options.http.startResumableUpload(
        token,
        folderId,
        file.name,
        file.mimeType,
        file.sizeBytes,
        folderHint,
      ),
    );
    let offset = 0;
    let finalMeta: DriveListMatch | null = null;
    for await (const rawChunk of file.stream()) {
      if (rawChunk.byteLength === 0) continue;
      let pos = 0;
      while (pos < rawChunk.byteLength) {
        const end = Math.min(pos + RESUMABLE_CHUNK_BYTES, rawChunk.byteLength);
        const slice = rawChunk.subarray(pos, end);
        // Retry the chunk PUT specifically; the session URI is stable
        // across retries per Drive's resumable protocol.
        finalMeta = await this.callWithRetry(() =>
          this.options.http.putChunk(session.sessionUri, slice, offset, file.sizeBytes),
        );
        offset += slice.byteLength;
        pos = end;
      }
    }
    if (offset !== file.sizeBytes) {
      throw new StorageException(
        StorageError.transport(
          `Resumable upload truncated: sent ${offset} of ${file.sizeBytes} bytes`,
        ),
      );
    }
    if (finalMeta === null) {
      throw new StorageException(
        StorageError.upstream(500, "Resumable upload completed without final metadata"),
      );
    }
    return matchToRemoteFile(finalMeta);
  }

  private async lookupOrCreateFolder(): Promise<string> {
    const token = await this.requireToken();
    const existing = await this.callWithRetry(() =>
      this.options.http.findAppFolder(token, null, APP_FOLDER_NAME),
    );
    if (existing !== null) return existing;
    return this.callWithRetry(() => this.options.http.createFolder(token, null, APP_FOLDER_NAME));
  }

  private async requireToken(): Promise<string> {
    const token = await this.options.token();
    if (token === null || token.length === 0) {
      throw new StorageException(
        StorageError.unauthorized("No Google Drive token available; ask the user to re-link"),
      );
    }
    return token;
  }

  private callWithRetry<T>(op: () => Promise<T>): Promise<T> {
    return retry(op, isRetriable, this.backoff);
  }
}

function matchToRemoteFile(m: DriveListMatch): RemoteFile {
  return {
    externalId: m.id,
    name: m.name,
    mimeType: m.mimeType,
    sizeBytes: m.size,
  };
}

function isRetriable(err: unknown): boolean {
  if (!(err instanceof StorageException)) return false;
  const inner = err.error;
  if (inner.kind === "TransportError") return true;
  if (inner.kind === "UpstreamError") {
    return inner.status === 408 || inner.status === 429 || inner.status >= 500;
  }
  return false;
}
